packages/flutter/lib/src/gestures/monodrag.dart

```diff
@@ -80,15 +87,31 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
     this.multitouchDragStrategy = MultitouchDragStrategy.latestPointer,
     this.velocityTrackerBuilder = _defaultBuilder,
     this.onlyAcceptDragOnThreshold = false,
+    this.cancelWhenOutsideBoundary = false,
+    this.createDragBoundary,
     super.supportedDevices,
     AllowedButtonsFilter? allowedButtonsFilter,
-  }) : super(allowedButtonsFilter: allowedButtonsFilter ?? _defaultButtonAcceptBehavior);
+  }) : assert(
+          !cancelWhenOutsideBoundary || createDragBoundary != null
+       ), super(allowedButtonsFilter: allowedButtonsFilter ?? _defaultButtonAcceptBehavior);
 
   static VelocityTracker _defaultBuilder(PointerEvent event) => VelocityTracker.withKind(event.kind);
 
   // Accept the input if, and only if, [kPrimaryButton] is pressed.
   static bool _defaultButtonAcceptBehavior(int buttons) => buttons == kPrimaryButton;
 
+  /// Called when the drag gesture starts. The returned [DragBoundary] will
+  /// be used to define the boundary for this drag gesture.
+  ///
+  /// See also:
+  /// * [DragBoundary], which defines the boundary of the drag gesture.
+  /// * [CreateDragBoundary], which is a callback that creates a [DragBoundary].
+  CreateDragBoundary? createDragBoundary;
```

Comment:
I'm wondering if this API is needed to accomplish the given use-cases. Are the use-cases achievable just using `DragRectBoundaryProvider`?

Here is the example without any API changes to `DragGestureRecognizer`.

```dart
class DragBoundaryExampleApp extends StatefulWidget {
  const DragBoundaryExampleApp({super.key});

  @override
  State<StatefulWidget> createState() => DragBoundaryExampleAppState();
}

class DragBoundaryExampleAppState extends State<DragBoundaryExampleApp> {
  Offset _basePosition = Offset.zero;
  Offset _dragPosition = Offset.zero;
  late DragBoundary? boundaryInfo;
  @override
  Widget build(BuildContext context) {
    final Offset offset = _dragPosition - _basePosition;
    return MaterialApp(
      home: Scaffold(
        body: Padding(
          padding: const EdgeInsets.all(100),
          child: DragRectBoundaryProvider(
            child: Stack(
              children: <Widget>[
                Container(
                  color: Colors.green,
                ),
                Positioned(
                  top: offset.dy,
                  left: offset.dx,
                  child: Builder(
                    builder: (BuildContext context) {
                      return GestureDetector(
                        behavior: HitTestBehavior.translucent,
                        onPanStart: (DragStartDetails details) {
                          boundaryInfo = DragRectBoundaryProvider.of(context)?.getDragBoundary(context, details.globalPosition);
                          setState(() {
                            _basePosition = details.globalPosition - offset;
                            _dragPosition = details.globalPosition;
                          });
                        },
                        onPanUpdate: (DragUpdateDetails details) {
                          setState(() {
                            if (boundaryInfo == null) {
                              return;
                            }
                            if (boundaryInfo!.isWithinBoundary(details.globalPosition)) {
                              _dragPosition = details.globalPosition;
                            } else {
                              _dragPosition = boundaryInfo!.getNearestPositionWithinBoundary(details.globalPosition);
                            }
                          });
                        },
                        child: Container(
                          width: 100,
                          height: 100,
                          color: Colors.red,
                        ),
                      );
                    },
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```

Reply:
This is great, but it's hard to cancel the gesture through the boundary without changing the API.

Reply:
I have some thoughts, as you said, we don't change any APIs to implement the boundary feature, and then let the user actively control the cancellation of the gesture in another way. I will work on this, if you have any ideas, please let me know. @Renzo-Olivares